pguijas/p2pfl
# Add Dockerfile

## Description

This pull request introduces a new Dockerfile designed to create an Ubuntu environment with Python. It includes the following key features:

1. An installation of the Python environment.
2. A shared folder (`/app`) between the Docker container and the host machine to facilitate easy access to project files.
3. Installation of all project dependencies using Poetry, a dependency management tool for Python.

This setup aims to streamline the development process by providing a consistent and reproducible environment for running and developing the project.

## Changes

- Added a Dockerfile to set up an Ubuntu-based container.
- Installed a Python environment within the container.
- Created a shared directory (`/app`) between the host and the container.
- Installed Poetry to handle project dependencies.
- Configured the container to install all dependencies specified in `pyproject.toml` using Poetry.
- Update the project's documentation to include instructions on using the new Docker setup.

Dockerfile based on [Building an Efficient Docker Image with Poetry and PyTorch](https://miguel-mendez-ai.com/2024/03/12/poetry-torch-docker)

## Todos

- Implement a CI/CD pipeline to build and test the Docker container.
- Add more detailed testing scripts to verify the container setup.
- Explore optimizations for reducing the Docker image size.

## How can we see the results or reproduce them?

To see the results or reproduce the environment setup:

1. Build the Docker image using the provided Dockerfile:
   ```sh
   cd p2pfl
   docker build -t p2pfl .
   ```

2. Run a container from the image
   ```sh
   docker run -it -p 8000:8000 --gpus=all -v .:/app/ p2pfl
   ```

## How Has This Been Tested?
- Unit test: Verified the installation of Python within the container.
- Integration test: Checked that the shared folder /app is correctly mounted and accessible.
- End-to-end test: Ensured that Poetry can install all dependencies from `pyproject.toml` without issues.
- System test: Ran the project within the container to confirm the environment is set up correctly. Tested in Docker 4.31.1.

## How do you want others to review your code?
- Entire diff
